metrics/metrics.go

```diff
@@ -92,28 +123,106 @@ func watchMetrics(
 	ticker.Start(false, true, true)
 }
 
-func watchRawMetrics(
-	client *client.Client,
-	source RawSource,
+func watchMetricsProm(
+	c *client.Client,
+	sources map[string]Source,
 	interval time.Duration,
 ) {
-	metricsPipe := make(chan RawMetrics)
-	go sendRawMetrics(client, metricsPipe)
-	defer close(metricsPipe)
 
-	ticker := utils.NewTicker("raw-metrics", interval, func() {
-		metrics, err := source.GetRawMetrics()
-		if err != nil {
-			client.Errorf(err, "unable to retrieve metrics from sink")
-		}
+	ticker := utils.NewTicker(
+		"prom-metrics",
+		interval,
+		func(tickTime time.Time) {
+			m := sync.Mutex{}
```

Comment:
create a context with deadline and pass it inside all external calls

Reply:
for http requests, calling `request = request.WithContext(ctx)` will make the request respect the deadline of the context

to create a context `ctx := context.WithTimeout(context.TODO(), time.Minute)`